app/middleware/user-is-authorised.js

```diff
@@ -0,0 +1,40 @@
+'use strict'
+
+const { NotAuthorisedError, NotAuthenticatedError, UserAccountDisabledError } = require('../errors')
+const { SERVICE_EXTERNAL_ID, GATEWAY_ACCOUNT_EXTERNAL_ID } = require('../paths').keys
+
+module.exports = function getUserIsAuthorisedMiddleware (permission) {
+  return function userIsAuthorised (req, res, next) {
+    const { user, session, params, service } = req
+
+    if (!user) {
+      return next(new NotAuthenticatedError('User not found on request'))
+    }
+    if (!session) {
+      return next(new NotAuthenticatedError('Session not found on request'))
+    }
+    if (user.sessionVersion !== session.version) {
+      return next(new NotAuthenticatedError(`Invalid session version - session version is ${session.version}, user session version is ${user.sessionVersion}`))
+    }
+    if (user.disabled) {
+      return next(new UserAccountDisabledError('User account is disabled'))
+    }
+    if (!(session.secondFactor && session.secondFactor === 'totp')) {
+      return next(new NotAuthenticatedError('Not completed second factor authentication'))
+    }
+
+    if (params[GATEWAY_ACCOUNT_EXTERNAL_ID] || params[SERVICE_EXTERNAL_ID]) {
+      if (!service) {
```

Comment:
I wonder if it would be good to add a sense check here that if there's a gateway account ID it belongs to the service that has been placed on the session. 

This currently relies on the request data middleware all behaving correctly, I can't think of a scenario where a session could be added that the gateway account doesn't belong to but it would be a fairly simple check and potentially help firm up edge cases.